compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -482,9 +482,48 @@ impl<'a> Parser<'a> {
     }
 
     /// Checks if this expression is a successfully parsed statement.
+    ///
+    /// This determines whether to continue parsing more of an expression in a
+    /// match arm (false) vs continue to the next arm (true).
+    ///
+    /// ```ignore (illustrative)
+    /// match ... {
+    ///     // Is this calling $e as a function, or is it the start of a new arm
+    ///     // with a tuple pattern?
+    ///     _ => $e (
+    ///             ^                                                          )
+    ///
+    ///     // Is this an Index operation, or new arm with a slice pattern?
+    ///     _ => $e [
+    ///             ^                                                          ]
+    ///
+    ///     // Is this a binary operator, or leading vert in a new arm? Same for
+    ///     // other punctuation which can either be a binary operator in
+    ///     // expression or unary operator in pattern, such as `&` and `-`.
+    ///     _ => $e |
+    ///             ^
+    /// }
+    /// ```
+    ///
+    /// If $e is something like `path::to` or `(…)`, continue parsing the same
+    /// arm.
+    ///
+    /// If $e is something like `{}` or `if … {}`, then terminate the current
+    /// arm and parse a new arm.
     fn expr_is_complete(&self, e: &Expr) -> bool {
+        // Surprising special case: even though braced macro calls like
+        // `m! {}` normally introduce a statement boundary when found at
+        // the head of a statement, in match arms they do not terminate
+        // the arm.
+        //
+        //     let _ = { m! {} () };  // macro call followed by unit
+        //
+        //     match ... {
+        //         _ => m! {} (),  // macro that expands to a function, which is then called
+        //     }
+        //
         self.restrictions.contains(Restrictions::STMT_EXPR)
-            && !classify::expr_requires_semi_to_be_stmt(e)
+            && !classify::expr_requires_comma_to_be_match_arm(e)
```

Comment:
Right, this is preserving old behavior. The exact same behavior that used to be incorrectly called `expr_requires_semi_to_be_stmt` is now called `expr_requires_comma_to_be_match_arm`.

Regarding naming, one possibility is that instead of distinguishing a "commas in match arms" and "semicolons in statements" case (and the implicit 3rd case, expressions that are neither statement nor match arm, such as a function arg), we could rename things along the following lines:

- "parse the longest possible expression"
- "parse an expression using earlier boundary rule" (for a match arm, range, or whatever else looks at expr_is_complete)
- "parse an expression until the earliest possible boundary" (for a statement)

Another possibility is to omit this specific commit (_"Add classify::expr_requires_comma_to_be_match_arm"_) and keep using `match e { MacCall(_) => …, _ => expr_requires_semi_to_be_stmt(e) }` in the cases where `expr_requires_semi_to_be_stmt` is not the desired behavior.

Reply:
One more option &mdash; instead of having what's currently in the PR:

```rust
// compiler/rustc_ast/src/util/classify.rs

pub fn expr_requires_semi_to_be_stmt(e: &ast::Expr) -> bool {
    match &e.kind {
        If(..)
        | Match(..)
        | Block(..)
        | While(..)
        | Loop(..)
        | ForLoop { .. }
        | TryBlock(..)
        | ConstBlock(..) => false,

        MacCall(mac_call) => mac_call.args.delim != Delimiter::Brace,

        _ => true,
    }
}

pub fn expr_requires_comma_to_be_match_arm(e: &ast::Expr) -> bool {
    match &e.kind {
        MacCall(_) => true,
        _ => expr_requires_semi_to_be_stmt(e),
    }
}
```

we could have:

```rust
pub fn expr_is_complete(e: &ast::Expr) -> bool {
    // this is negative of the old "expr_requires_semi_to_be_stmt"
    match &e.kind {
        If(..)
        | Match(..)
        | Block(..)
        | While(..)
        | Loop(..)
        | ForLoop { .. }
        | TryBlock(..)
        | ConstBlock(..) => true,

        MacCall(_) | _ => false,
    }
}

pub fn expr_requires_semi_to_be_stmt(e: &ast::Expr) -> bool {
    match &e.kind {
        MacCall(mac_call) => mac_call.args.delim != Delimiter::Brace,
        _ => !expr_is_complete(e),
    }
}
```

Reply:
I prefer either the final option or the renaming you suggested first above. I'll leave it up to you. Otherwise this PR looks good to me.

Reply:
I removed https://github.com/rust-lang/rust/commit/5d5954c74df2b6fdfc89a29b502bb898a51fcd1d and added https://github.com/rust-lang/rust/commit/f895205d41ced8ddc7b9f72fdf9e474292aa2cb8.